File: 107. Binary Tree Level Order Traversal II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    if ( !root ) {
        return [];
    }

    var result = [];
    var queue = [];
    queue.push( root );

    while ( queue.length > 0 ) {
        var length = queue.length;
        var level = [];

        for ( var i = 0; i < length; i++ ) {
            var node = queue.shift();
            level.push( node.val );
            if ( node.left ) {
                queue.push ( node.left );
            }
            if ( node.right ) {
                queue.push ( node.right );
            }
        }

        result.unshift( level );
    }

    return result;
};

File: 107. Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque


class Solution(object):
    def levelOrderBottom(self, root):
        result = deque()
        if root is None:
            return list(result)
        q = deque()
        q.append(root)
        level = 1

        while q:
            l = len(q)
            tmp = []
            for i in xrange(l):
                node = q.popleft()
                tmp.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level += 1
            result.appendleft(tmp)
        return list(result)
